### ISSUE TITLE
Inserted keyframes in material node graph don't show up in graph editor or anywhere else

### ISSUE THREAD

**Issue by novacrazy** [LuxCoreRender/BlendLuxCore#167]
#### Description

I've been trying to animate some material properties from the node graph, and while I can add keyframes to node fields (like animating opacity), they simply don't show up in the graph editor or dope sheet like they would for Cycles and so forth.

Unfortunately, this also forces the default (non-linear) interpolation between those keyframes.

#### System Information

Windows 10 Pro with GTX 970 GPU

#### Software Version

* Blender version: 2.79b
* LuxCore version: 2.1alpha0

Related: 
* https://blender.stackexchange.com/questions/94329/how-do-i-view-animation-from-an-animation-nodes-node-tree
* https://github.com/JacquesLucke/animation_nodes/issues/820
* https://github.com/JacquesLucke/animation_nodes/issues/359 (this one contains an interesting quote from a Blender developer)

**Comment by Theverat** [comment]
Unfortunately, this is a Blender limitation.
All addons with custom nodes are affected (e.g. Sverchok, Animation Nodes).

**Comment by novacrazy** [comment]
Dang, that's a bummer. Has there been any activity in Blender 2.8 that would allow for third-party node animations?

**Comment by Theverat** [comment]
I haven't seen any, but I don't follow the development too closely.
You could try to ask in their developer IRC or so.

**Comment by Theverat** [comment]
What you can do if you know a little Python is something like the following:

![scrn_2018-06-02_11-06-06](https://user-images.githubusercontent.com/1584761/40872818-44d52624-6655-11e8-8129-55f10b85165c.png)
```python
import bpy
from bpy.app.handlers import persistent


@persistent
def handler(scene):
    node_tree = bpy.data.node_groups["Material.004"]
    matte = node_tree.nodes["Matte Material"]
    
    frame = scene.frame_current
    matte.inputs["Opacity"].default_value = frame / 100


def register():
    bpy.app.handlers.frame_change_post.append(handler)
    
def unregister():
    bpy.app.handlers.frame_change_post.remove(handler)

if __name__ == "__main__":
    register()
```

Another idea is: I could create a node that outputs the current frame, and you can then work with this with the math nodes (however they are a bit limited at the moment).
This node could also be more general: allowing access to any property of any datablock.

**Comment by novacrazy** [comment]
That’s way too hardcoded for my taste. I’d prefer to be able to control values via the graph editor ideally. 

A node which could select an object (like how the smoke data node allows you to select the domain), and extract transform values would be a functional workaround to animate arbitrary properties by animating those transforms.

**Comment by Theverat** [comment]
> A node which could select an object (like how the smoke data node allows you to select the domain), and extract transform values would be a functional workaround to animate arbitrary properties by animating those transforms.

Yes, this is possible with the "more general" version of the node I mentioned: 

> This node could also be more general: allowing access to any property of any datablock.

The datablock could be an object, a scene, image etc.

**Comment by Theverat** [comment]
I have an initial draft working, however it still needs some refinement.
For example, it seems like PointerProperty only allows to store one kind of subclass of ID: https://docs.blender.org/api/2.79/bpy.types.ID.html#bpy.types.ID
Not any subclass. So currently you can only select objects.
I'll see if I can find a solution that allows any ID type.

Another thing I'd like to support is accessing attributes through multiply calls, e.g. `object.location.x`.

![scrn_2018-06-02_12-26-03](https://user-images.githubusercontent.com/1584761/40873455-9260c118-6660-11e8-95d4-ea422ed2e776.png)
![scrn_2018-06-02_12-26-28](https://user-images.githubusercontent.com/1584761/40873456-927f4750-6660-11e8-8dc1-51488a693b3b.png)

**Comment by Theverat** [comment]
You can do funny stuff with this node.
Like drive the color of a material with the color of the node itself.

![scrn_2018-06-02_18-18-57](https://user-images.githubusercontent.com/1584761/40878442-87e626c0-6691-11e8-82a3-2cbcc2ca820b.png)

**Comment by Theverat** [comment]
Another idea: show a dropdown with the possible attributes.
Maybe even filter for actually usable ones, like scalars, vectors, colors etc.

![scrn_2018-06-02_19-52-39](https://user-images.githubusercontent.com/1584761/40879349-bbd560b0-669e-11e8-9a05-314c0c5c6f18.png)

**Comment by Theverat** [comment]
> Another idea: show a dropdown with the possible attributes.
Maybe even filter for actually usable ones, like scalars, vectors, colors etc.

This turned out to be much more complicated than I thought, so I dropped this idea.
However I'm still considering to add a popup that shows all available attributes of a datablock (but only the first level).

Stuff I added:
* Any ID type, not just object: you can now select the datablock type from a dropdown
* Indexed access of collections (slicing is also supported, but you can't leave out parameters)
* Auto-detection of the correct socket type (value or color) to help a bit, errors are now also shown after entering the string, not only if you export

![scrn_2018-06-04_19-51-36](https://user-images.githubusercontent.com/1584761/40933117-d67c02b4-6830-11e8-802a-33d9f6792c66.png)

**Comment by novacrazy** [comment]
On your checklist, for the function call example, it would probably be better and more flexible for artists to add those as other nodes like I listed in LuxCoreRender/LuxCore#118.

A node for vector normalization and another for splitting vectors/colors into their scalar components would reduce the complexity of the property access node, and add extra flexibility for other stuff like what I talked about in the other issue.

Also, even though this technically started as a workaround, I’m very excited for this now. Dynamic effects based on other object arbitrary properties is exactly what every offline renderer needs, in my opinion, and by wrapping it in even the most barebones of nodes gives a lot of power to artists.

Even now, I just had an idea to use this to compute distance from an object inside of a heterogenous volume to add procedural clouds or fog around that object.

**Comment by Theverat** [comment]
Yes, I think I will not add function call support for now because it adds another bunch of complexity. For example, how to handle functions that expect arguments? Maybe this can be added at a later time.

The only thing I want to try is to add a popup that shows the available attributes (basically a filtered `dir()`), so you don't have to check in the Python console every time.

**Comment by novacrazy** [comment]
One idea I had for easier deeply nested property access, at the cost of another new node and a connection data type, was to separate out the object selection node and a property access node, then have all the property access nodes work recursively (sort of). Additionally, an array index/slice menu on the property node.

So for your `material_slots[0].material.diffuse_color[0:2]` example if would become:

```
Cube:                   Object Select Node -> 
material_slots[0]:      Property Node (with Array Index) -> 
material:               Property Node -> 
diffuse_color[0:2]:     Property Node (with Array Slice) -> Color Output
```

On each Property Node it would auto-populate a dropdown with `dir()`. If the selected property is a scalar/vector/color, it could output it there. If the selected property is another nested object, it would output that and sub-properties could be selected with another Property Node. 

If the selected property is an array, it would offer the index/slice options and do the same as the previous two outcomes with the result of that index/slice operation.

Furthermore, this would allow for branching selections, like if they want to select the `diffuse_color` and `specular_color` properties from the same object material slot, they could branch off of the same parent Property Node. Or if they want to select multiple transforms from the same object, it would reduce repetition by only selecting the object once in the Object Select Node.

What do you think of that?

**Comment by Theverat** [comment]
This sounds good, it solves several issues in an elegant way.
It also adheres to the "UI principle of discoverability" in a better way than my node, because the dir() dropdown, output socket and array/slice option show the user all he needs to explore.
With this proposal there is probably no need to keep the Python console open at the side.

Do you want to have a go at it?
My time is currently rather short, unfortunately.

**Comment by novacrazy** [comment]
Well, it's a start:
![selectobject](https://user-images.githubusercontent.com/4708645/41080137-9e55721a-69ea-11e8-9f3f-b75d26a60844.PNG)
And some minimal sorting of the data types
![sorted data types](https://user-images.githubusercontent.com/4708645/41079806-fe39c52a-69e8-11e8-97ef-f7344c5ab659.PNG)

So I decided I would like to give this a try. I've never worked with Blender's API before, and it's been a year or two since I've touched Python, so it took a few hours to figure out things like the custom socket types and nodes in general. I suppose next up is tracing back through the tree to get the correct properties and dynamically generate the selection options.

**Comment by Theverat** [comment]
Looking good.
If you have questions, feel free to ask me.
And if you want me to review code, post a link and I'll have a look.

**Comment by novacrazy** [comment]
I now more or less have the recursion done, and I was able to get reroute nodes working.

![selectobject2](https://user-images.githubusercontent.com/4708645/41133813-af35d3c8-6a8e-11e8-83a7-3bb01a7962e4.PNG)

Next up is list slicing and correct output types, and then finally getting it to export these to LuxCore texture constants.

**Comment by canonex** [comment]
Alternative coded solution is to directly apply values from driver function:
https://forums.luxcorerender.org/viewtopic.php?f=4&t=626#p6229

Not a perfect solution but it works.

**Comment by LaurieAnnis** [comment]
Apparently you have to have the object AND the node selected, for the keyframes to appear.  I found this old thread because I was having the same issue, and stumbled across this solution on another site, and it worked for me.

**Comment by Theverat** [comment]
> Apparently you have to have the object AND the node selected, for the keyframes to appear.

This is the case for Cycles nodes, but it does not work on LuxCore node trees - the keyframes never appear in the graph editor.

**Comment by NazzarenoGiannelli** [comment]
I am having the same issue with Sverchok. I am trying to simply keyframe a Vector Lerp, but keyframes do not show up and values I modify get not registered.
Anyone knows how to solve this?

**Comment by Viper1-1** [comment]
Alright, I have found a solution!
<img width="1680" alt="Screen Shot 2022-09-30 at 10 07 48 PM" src="https://user-images.githubusercontent.com/102196609/193378941-9c9984f5-d5de-413c-bc32-7a04cc611f5a.png">
<img width="1680" alt="Screen Shot 2022-09-30 at 10 07 55 PM" src="https://user-images.githubusercontent.com/102196609/193378944-1e3ec88c-ea06-44a4-9bdc-7e812eca3f93.png">

**Comment by Viper1-1** [comment]
First select the node with the keyframes, go into the Dope Sheet, and voila, there is the keyframe! and you can edit it and everything!
